pkg/aws/vpcendpoint/client_update.go

```diff
@@ -0,0 +1,132 @@
+package vpcendpoint
+
+import (
+	"context"
+	"sort"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/giantswarm/microerror"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/giantswarm/aws-vpc-operator/pkg/errors"
+)
+
+type UpdateVpcEndpointInput struct {
+	RoleARN                 string
+	Region                  string
+	VpcEndpointId           string
+	CurrentSubnetIds        []string
+	CurrentSecurityGroupIds []string
+	WantedSubnetIds         []string
+	WantedSecurityGroupIds  []string
+	Tags                    map[string]string
+}
+
+func (c *client) Update(ctx context.Context, input UpdateVpcEndpointInput) (err error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Started updating VPC endpoint")
+	defer func() {
+		if err == nil {
+			logger.Info("Finished updating VPC endpoint")
+		} else {
+			logger.Error(err, "Failed to update VPC endpoint")
+		}
+	}()
+
+	if input.RoleARN == "" {
+		return microerror.Maskf(errors.InvalidConfigError, "%T.RoleARN must not be empty", input)
+	}
+	if input.Region == "" {
+		return microerror.Maskf(errors.InvalidConfigError, "%T.Region must not be empty", input)
+	}
+	if input.VpcEndpointId == "" {
+		return microerror.Maskf(errors.InvalidConfigError, "%T.VpcId must not be empty", input)
+	}
+	if len(input.WantedSubnetIds) == 0 {
+		return microerror.Maskf(errors.InvalidConfigError, "%T.WantedSubnetIds must not be empty", input)
+	}
+	if len(input.WantedSecurityGroupIds) == 0 {
+		return microerror.Maskf(errors.InvalidConfigError, "%T.WantedSecurityGroupIds must not be empty", input)
+	}
+
+	// Sort current securityGroupIds and subnetIds, so we can use sort.SearchStrings
+	// when checking difference in slices.
+	// This modifies slice in-place, but we just use it here anyway, so that's
+	// fine.
+	sort.Strings(input.CurrentSecurityGroupIds)
+	sort.Strings(input.CurrentSubnetIds)
+
+	// securityGroupIDs that we will add, those specified in the input, but not
+	// already present in current state
+	securityGroupIdsToBeAdded := diff(input.WantedSecurityGroupIds, input.CurrentSecurityGroupIds)
```

Comment:
I'll move checking of difference to outside of client method. Wanted to make it more declarative (as opposed to AWS's add/remove operations), but it doesn't look very nice TBH.